app/main/login.py

```diff
@@ -60,24 +64,69 @@ def forgotten_password():
 
 @main.route('/forgotten-password', methods=["POST"])
 def send_reset_email():
-    return redirect(url_for('.change_password'))
-
-
-@main.route('/change-password', methods=["GET"])
-@login_required
-def change_password():
+    email_address = request.form['email-address']
+    user = api_client.user_by_email(email_address)
+    if user is not None:
+        # Send a password reset email with token
+        current_app.logger.info(
+            "Sending password reset email for supplier %d (%s)",
+            user.id, user.email_address
+        )
+        email.send_password_email(user.id, email_address)
+        # TODO: Add to count in "forgotten password emails sent" metric
+    else:
+        current_app.logger.info(
+            "Password reset request for invalid supplier email address %s",
+            email_address
+        )
+        # TODO: Add to count in "forgotten password - invalid" metric
+
+    flash('If that Digital Marketplace supplier account exists, you will '
+          'be sent an email containing a link to reset your password.')
+    return redirect(url_for('.forgotten_password'))
+
+
+@main.route('/change-password/<token>', methods=["GET"])
+def change_password(token):
+    try:
+        decoded = email.decode_email(token)
+        user_id = decoded["user"]
+        email_address = decoded["email"]
+    except SignatureExpired:
+        current_app.logger.info("Password reset attempt with expired token.")
+        flash('The token supplied has expired. Password reset links are only'
+              ' valid for 24 hours. You can generate a new one using the form'
+              ' below.', 'error')
+        return redirect(url_for('.forgotten_password'))
+    except BadSignature as e:
+        current_app.logger.info("Error changing password: %s", e)
+        flash('The token supplied was invalid. You can generate a new one'
```

Comment:
Not sure about having flash messages hard coded in python code. Do we need a messages file or some other model?